### PR TITLE
Give std/var correction overloads proper defaults

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #55680
* #55679
* #87118
* __->__ #56398

The correction overloads defaults were left off for forward
compatibility reasons, but this FC window expired well over a year ago
at this point.

Differential Revision: [D29625593](https://our.internmc.facebook.com/intern/diff/D29625593)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @lezcano @fdrocha

### CODE REVIEW

**Comment by peterbell10** [comment]
@mruberry PTAL, forward-compat period should be over now.

File: torch/_tensor_docs.py [hunk 1]
```diff
@@ -3570,7 +3570,12 @@ def callable(a, b) -> number
 
 add_docstr_all('std',
                r"""
-std(dim, unbiased=True, keepdim=False) -> Tensor
+std(dim=None, *, correction=1, keepdim=False) -> Tensor
+
+See :func:`torch.std`
+
+.. function:: std(dim, unbiased=True, keepdim=False) -> Tensor
```
Comment (mruberry) [inline torch/_tensor_docs.py:3577]: Let's just remove these legacy signatures from the docs

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
...
+    ...      [ 1.5027, -0.3270,  0.5905,  0.6538],
+    ...      [-1.5745,  1.3330, -0.5596, -0.6548],
+    ...      [ 0.1264, -0.5080,  1.6420,  0.1992]])
+    >>> torch.std(a, dim=1, keepdim=True)
+    tensor([[1.0311],
+            [0.7477],
+            [1.2204],
+            [0.9087]])
+
+.. function:: std(input, dim, unbiased, keepdim=False, *, out=None) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:8504]: Let's kill these definitions, too

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -9724,7 +9808,48 @@ def merge_dicts(*dicts):
 """.format(**common_args))
 
 add_docstr(torch.var, r"""
-var(input, dim, unbiased, keepdim=False, *, out=None) -> Tensor
+var(input, dim=None, *, correction=1, keepdim=False, out=None) -> Tensor
+
+Calcualtes the variance over the dimensions specified by :attr:`dim`. :attr:`dim`
```
Comment (mruberry) [inline torch/_torch_docs.py:9813]: "Calcualtes" -> "Calculates"

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
...
+
+.. math:: \sigma = \sqrt{\frac{1}{N - \delta N}\sum_{i=0}^{N-1}(x_i-\bar{x})^2}
+
+where :math:`x` is the sample set of elements, :math:`\bar{x}` is the
+sample mean, :math:`N` is the number of samples and :math:`\delta N` is
+the :attr:`correction`.
+""" + r"""
+
+{keepdim_details}
+
```
Comment (mruberry) [inline torch/_torch_docs.py:8480]: Let's add a note here (and to the other functions about the historic behavior):

"In PyTorch 1.9 and earlier _std_ accepted an _unbiased_ argument. If _unbiased_ was True (the default) it implied `correction=1` (the current default). If _unbiased_ was False then it implied `correction=0`. The new _correction_ parameter generalizes the deprecated _unbiased_ parameter."

How does that sound?

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -8461,7 +8461,48 @@ def merge_dicts(*dicts):
...
+Args:
+    {input}
+    {dim}
+
+Keyword args:
+    correction (int): difference between the sample size and sample degrees of freedom.
+                      Defaults to Bessel's correction, ``correction = 1``.
```
Comment (mruberry) [inline torch/_torch_docs.py:8487]: Is it worth mentioning that when correction = 1 this is an unbiased estimator of the standard deviation?

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -9724,7 +9808,48 @@ def merge_dicts(*dicts):
...
```
Comment (mruberry) [inline torch/_torch_docs.py:9834]: Is it worth noting that when correction = 1 this is an unbiased estimator of the population variance?

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -2673,7 +2673,7 @@ def sample_inputs_std_var(op_info, device, dtype, requires_grad, **kwargs):
         SampleInput(tensor_1d, kwargs=dict(dim=0, unbiased=False, keepdim=False)),
 
         SampleInput(tensor_nd, kwargs=dict(dim=(1,), correction=S // 2)),
-        SampleInput(tensor_nd, kwargs=dict(dim=None, correction=0, keepdim=True)),
+        SampleInput(tensor_nd, kwargs=dict(correction=0, keepdim=True)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2676]: Let's change the above "unbiased" uses to use "correction" and add a SampleInput with a correction factor that is not 0 or 1

**Review by mruberry — COMMENTED** [review]
Thanks for the ping, @peterbell10!

This looks really good. I made a few comments inline. Basically I want to hide the old parameter, add a note explaining that it existed and how people should update to the new functionality, and consider elaborating on the mathematical impact of the correction factor.

Looking forward to hearing your thoughts.

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:2970]: >Let's change the above "unbiased" uses to use "correction" 

The unbiased overloads are needed for backwards-compatibility so IMO they should be tested.

>add a SampleInput with a correction factor that is not 0 or 1

The line above this has `correction=S//2` which is 2.

File: torch/_torch_docs.py (hunk 5 above)
Comment (peterbell10) [reply torch/_torch_docs.py]: I've added a link to the wiki for Bessel's correction. IMO It's a bit more complicated than just being unbiased, it would have to be more like 
> assuming the input is a sample from a normally-distributed population, this gives an unbiased estimator of the population standard deviation.

which is probably too wordy for an argument description.

File: torch/_torch_docs.py (hunk 4 above)
Comment (peterbell10) [reply torch/_torch_docs.py:8474]: Tweaked the wording slightly to:
>   In PyTorch 1.8 and earlier `std` accepted an `unbiased` argument which was a less general form of the `correction` parameter. If `unbiased` was ``True`` it implied ``correction=1`` (this was the default). If `unbiased` was ``False`` then it implied ``correction=0``.

**Comment by peterbell10** [comment]
@mruberry PTAL

File: test/test_namedtensor.py [hunk 8]
```diff
@@ -1241,26 +1245,26 @@ def kthvalue_wrapper(tensor, *args, **kwargs):
...
-            Case(torch.var_mean, True, True, False, True, None),
-            Case(torch.min, True, False, True, True, values_and_indices),
-            Case(torch.max, True, False, True, True, values_and_indices),
-            Case(torch.unbind, False, False, False, False, None),
-            Case(torch.logsumexp, False, True, True, True, None),
-            Case(torch.mode, False, False, True, True, values_and_indices),
-            Case(kthvalue_wrapper, False, False, True, True, values_and_indices),
-            Case(torch.median, True, False, True, True, values_and_indices),
-            Case(torch.nanmedian, True, False, True, True, values_and_indices),
+            Case(torch.sum, False, True, True, True, True, None),
```
Comment (mruberry) [inline test/test_namedtensor.py:1248]: fyi @heitorschueroff; it'll be nice to revise this test so we don't deal with all these unnamed tuples

File: torch/_tensor_docs.py (hunk 1 above)
Comment (mruberry) [reply torch/_tensor_docs.py:3577]: Can we remove these or were you thinking we should keep them?

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -8448,81 +8448,110 @@ def merge_dicts(*dicts):
...
+The standard deviation (:math:`\sigma`) is calculated as
+
+.. math:: \sigma = \sqrt{\frac{1}{N - \delta N}\sum_{i=0}^{N-1}(x_i-\bar{x})^2}
+
+where :math:`x` is the sample set of elements, :math:`\bar{x}` is the
+sample mean, :math:`N` is the number of samples and :math:`\delta N` is
+the :attr:`correction`.
+""" + r"""
+
+Note:
```
Comment (mruberry) [inline torch/_torch_docs.py:8466]: Nice note.

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -9711,80 +9740,109 @@ def merge_dicts(*dicts):
...
+                      Defaults to `Bessel's correction`_, ``correction=1``.
     {keepdim}
     {out}
 
 Returns:
-    A tuple (var, mean) containing the variance and mean.
-
-.. function:: var_mean(input, unbiased) -> (Tensor, Tensor)
-   :noindex:
+    A tuple (var, mean) containing the variance and mean. Here 2
```
Comment (mruberry) [inline torch/_torch_docs.py:9831]: I think the "Here 2" is extraneous?

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:2970]: Cool

**Review by mruberry — COMMENTED** [review]
Looks really good, @peterbell10, just a few small comments inline

File: torch/_torch_docs.py (hunk 10 above)
Comment (peterbell10) [reply torch/_torch_docs.py]: Thanks, that was left over from testing that the docs were rendering properly. Removed.

**Comment by mruberry** [comment]
Everything looks good, @peterbell10. Last question is if this should remove the documentation of the `unbiased` signatures in torch/_tensor_docs.py? I think so, right?

**Comment by peterbell10** [comment]
True, it should match the function docs. Fixed.

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @peterbell10!

File: test/backward_compatibility/check_backward_compatibility.py [hunk 11]
```diff
@@ -79,6 +79,10 @@
     ("aten::divide", datetime.date(2021, 4, 28)),
     ("aten::_var", datetime.date(2021, 5, 28)),
     ("aten::_std", datetime.date(2021, 5, 28)),
+    ("aten::std", datetime.date(2021, 5, 28)),
```
Comment (mruberry) [inline test/backward_compatibility/check_backward_compatibility.py:82]: Last issue: these entries need to be updated since the old dates expired:

```
Jul 07 19:50:17 Broken ops: [
Jul 07 19:50:17 	aten::std_mean(Tensor self, bool unbiased=True) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std_mean.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std_mean.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::std(Tensor self, bool unbiased=True) -> (Tensor)
Jul 07 19:50:17 	aten::std.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::std.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::std.names_out(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::std.out(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var(Tensor self, bool unbiased=True) -> (Tensor)
Jul 07 19:50:17 	aten::var.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::var.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor)
Jul 07 19:50:17 	aten::var.names_out(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var.out(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False, *, Tensor(a!) out) -> (Tensor(a!))
Jul 07 19:50:17 	aten::var_mean(Tensor self, bool unbiased=True) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::var_mean.dim(Tensor self, int[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
Jul 07 19:50:17 	aten::var_mean.names_dim(Tensor self, str[1] dim, bool unbiased=True, bool keepdim=False) -> (Tensor, Tensor)
```

**Comment by mruberry** [comment]
Sorry @peterbell10, one last update. I'll start testing this internally now

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

File: test/backward_compatibility/check_backward_compatibility.py (hunk 11 above)
Comment (peterbell10) [reply test/backward_compatibility/check_backward_compatibility.py]: Ah, right. I've updated the date and rebased on a newer master.

**Comment by mruberry** [comment]
I think the date has to be valid:

```
Jul 09 01:15:08   File "check_backward_compatibility.py", line 82, in <module>
Jul 09 01:15:08     ("aten::std", datetime.date(2021, 8, 0)),
Jul 09 01:15:08 ValueError: day is out of range for month
```

**Comment by peterbell10** [comment]
Dates are fixed now and the backward compatibility check is passing.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

**Comment by peterbell10** [comment]
@mruberry any update on this PR? Were there issue with FC/BC that need to be fixed?

**Comment by mruberry** [comment]
> @mruberry any update on this PR? Were there issue with FC/BC that need to be fixed?

No update yet on this PR yet, but I'm working with our compiler and mobile teams now to review this and other BC/FC issues. I hope we'll have something sorted soonish.

**Comment by lezcano** [comment]
What's happened to the compiler @mruberry?

**Comment by peterbell10** [comment]
@mruberry PTAL, this shouldn't have any BC implications any more.

File: torch/_torch_docs.py (hunk 9 above)
Comment (mruberry) [reply torch/_torch_docs.py:10628]: Should this be converted to a version changed directive now? Also, it would be PyTorch 1.13?

File: torch/_torch_docs.py [hunk 12]
```diff
@@ -10649,45 +10667,58 @@ def merge_dicts(*dicts):
...
+.. math:: \sigma = \sqrt{\frac{1}{N - \delta N}\sum_{i=0}^{N-1}(x_i-\bar{x})^2}
+
+where :math:`x` is the sample set of elements, :math:`\bar{x}` is the
+sample mean, :math:`N` is the number of samples and :math:`\delta N` is
+the :attr:`correction`.
+
+"""
+    + r"""
+
+Note:
```
Comment (mruberry) [inline torch/_torch_docs.py:10687]: Same question about the note here

**Review by mruberry — APPROVED** [review]
Looks good -- see suggestion to modernize the documentation slightly. It's unfortunately this causes some `out` tests to fail now -- we should probably get that fixed

**Comment by peterbell10** [comment]
The `test_out` issue is that the `var(Tensor self, bool unbiased=True)` overload doesn't support an out argument, but the other overloads do. I don't really want to add out support to the deprecated overload, but I suppose we could just make that overload its own opinfo variant?

**Comment by peterbell10** [comment]
By the way, whats the best way to deal with the "Meta Internal-Only Changes Check". I assume that's no longer relevant so can the PR  be unimported, or would force-merging work be safe, or does this have to be landed internally now.

**Comment by mruberry** [comment]
> The `test_out` issue is that the `var(Tensor self, bool unbiased=True)` overload doesn't support an out argument, but the other overloads do. I don't really want to add out support to the deprecated overload, but I suppose we could just make that overload its own opinfo variant?

Yes, that would work. There are other operators with multiple OpInfos for similar reasons.

**Comment by mruberry** [comment]
> By the way, whats the best way to deal with the "Meta Internal-Only Changes Check". I assume that's no longer relevant so can the PR be unimported, would force-merging be safe, or does this have to be landed internally now?



Good question. @malfet @seemethere, what do you recommend?

**Comment by peterbell10** [comment]
@malfet @seemethere ping

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

**Comment by peterbell10** [comment]
@malfet were there any mobile issues?

**Comment by lezcano** [comment]
@malfet what's the state of this PR? Can we land it?

**Comment by malfet** [comment]
Sorry for the late reply, but can you please rebase again the latest PyTorch and I'll try to import it again (as you can see it can not be imported as is)

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

**Comment by malfet** [comment]
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29625593).

**Comment by malfet** [comment]
Ok, just landed internally, I wonder if it will be auto-closed by the bot now...

File: torch/testing/_internal/common_methods_invocations.py [hunk 13]
```diff
@@ -4960,10 +4960,16 @@ def sample_inputs_std_var(op_info, device, dtype, requires_grad, **kwargs):
     yield SampleInput(tensor_nd(), dim=(1,), correction=S // 2)
     yield SampleInput(tensor_nd(), dim=None, correction=0, keepdim=True)
     yield SampleInput(tensor_nd(), dim=None, correction=None)
+    yield SampleInput(tensor_nd(), correction=0, keepdim=True)
+
+
+def sample_inputs_std_var_unbiased(op_info, device, dtype, requires_grad, **kwargs):
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:4966]: why do we need unbiased variants of OpInfos? Can't we add them to the regular sample inputs function?

Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:4966]: The `unbiased` variant has `supports_out=False` due to some overloads not having an equivalent `out` overload, while `correction` overloads supports the `out` argument on all signatures.
